File: SundeeFundee/Sources/SundeeFundeeKit/DomainLayer/Challenge/ChallengeEngine.swift
import Foundation

// MARK: - ChallengeEngine

/// Pure domain logic for weight volume challenges.
///
/// All functions are static and side-effect-free — they accept data
/// and return data without touching any persistence layer.
public enum ChallengeEngine {

    // MARK: - Default Tiers

    /// Standard lifetime challenge tiers: Bronze (250K), Silver (500K), Gold (1M).
    public static func defaultLifetimeTiers() -> [ChallengeTier] {
        [
            ChallengeTier(name: "Bronze", targetVolumeLbs: 250_000, ordinal: 0),
            ChallengeTier(name: "Silver", targetVolumeLbs: 500_000, ordinal: 1),
            ChallengeTier(name: "Gold", targetVolumeLbs: 1_000_000, ordinal: 2),
        ]
    }

    // MARK: - Challenge Creation

    /// Creates a lifetime volume challenge with default tiers.
    public static func createLifetimeChallenge() -> Challenge {
        Challenge(
            type: .lifetimeVolume,
            title: "Lifetime Volume",
            tiers: defaultLifetimeTiers()
        )
    }

    /// Creates a per-exercise volume challenge.
    public static func createExerciseChallenge(
        exerciseName: String,
        tiers: [ChallengeTier]
    ) -> Challenge {
        Challenge(
            type: .exerciseVolume,
            title: "\(exerciseName) Volume",
            exerciseName: exerciseName,
            tiers: tiers
        )
    }

    /// Creates a custom time-boxed challenge.
    public static func createCustomChallenge(
        title: String,
        targetVolumeLbs: Double,
        endDate: Date? = nil
    ) -> Challenge {
        let tier = ChallengeTier(name: "Goal", targetVolumeLbs: targetVolumeLbs, ordinal: 0)
        return Challenge(
            type: .custom,
            title: title,
            tiers: [tier],
            endDate: endDate
        )
    }

    // MARK: - Progress Computation

    /// Computes the current progress of a challenge.
    public static func computeProgress(challenge: Challenge) -> ChallengeProgress {
        let tiers = challenge.tiers.sorted { $0.ordinal < $1.ordinal }

        guard !tiers.isEmpty else {
            return ChallengeProgress(
                currentTierName: "—",
                percentComplete: 0,
                volumeRemaining: 0,
                isFullyComplete: true
            )
        }

        // If all tiers completed
        if challenge.currentTierIndex >= tiers.count || challenge.status == .completed {
            let lastTier = tiers.last!
            return ChallengeProgress(
                currentTierName: lastTier.name,
                percentComplete: 1.0,
                volumeRemaining: 0,
                isFullyComplete: true
            )
        }

        let currentTier = tiers[challenge.currentTierIndex]
        let previousTarget = challenge.currentTierIndex > 0
            ? tiers[challenge.currentTierIndex - 1].targetVolumeLbs
            : 0

        let tierRange = currentTier.targetVolumeLbs - previousTarget
        let tierProgress = challenge.accumulatedVolumeLbs - previousTarget
        let percent = tierRange > 0 ? min(max(tierProgress / tierRange, 0), 1.0) : 1.0
        let remaining = max(currentTier.targetVolumeLbs - challenge.accumulatedVolumeLbs, 0)

        return ChallengeProgress(
            currentTierName: currentTier.name,
            percentComplete: percent,
            volumeRemaining: remaining,
            isFullyComplete: false
        )
    }

    // MARK: - Volume Update

    /// Updates a challenge with additional volume, advancing tiers as needed.
    ///
    /// - Returns: Updated challenge. Check the returned `ChallengeProgress` for `justCompletedTier`.
    public static func updateVolume(
        challenge: Challenge,
        additionalVolumeLbs: Double
    ) -> (challenge: Challenge, progress: ChallengeProgress) {
        var updated = challenge
        updated.accumulatedVolumeLbs += additionalVolumeLbs

        let tiers = updated.tiers.sorted { $0.ordinal < $1.ordinal }
        var justCompleted: ChallengeTier?

        // Advance through tiers
        while updated.currentTierIndex < tiers.count {
            let tier = tiers[updated.currentTierIndex]
            if updated.accumulatedVolumeLbs >= tier.targetVolumeLbs {
                justCompleted = tier
                updated.currentTierIndex += 1
            } else {
                break
            }
        }

        // Check if fully complete
        if updated.currentTierIndex >= tiers.count {
            updated.status = .completed
        }

        var progress = computeProgress(challenge: updated)
        progress = ChallengeProgress(
            currentTierName: progress.currentTierName,
            percentComplete: progress.percentComplete,
            volumeRemaining: progress.volumeRemaining,
            justCompletedTier: justCompleted,
            isFullyComplete: progress.isFullyComplete
        )

        return (updated, progress)
    }

    // MARK: - Retroactive Volume

    /// Computes total historical volume from workout history.
    ///
    /// - Parameters:
    ///   - workouts: All completed workouts.
    ///   - exerciseName: If provided, only counts volume from this exercise.
    /// - Returns: Total volume in pounds.
    public static func retroactiveVolume(
        from workouts: [Workout],
        exerciseName: String? = nil
    ) -> Double {
        var total = 0.0
        for workout in workouts where workout.isComplete {
            for exercise in workout.exercises {
                if let name = exerciseName, exercise.name != name {
                    continue
                }
                for set in exercise.targetSets {
                    let reps = set.actualReps ?? set.reps
                    let weight = (set.completedWeight ?? 0) > 0
                        ? set.completedWeight!
                        : (set.prescribedWeight > 0 ? set.prescribedWeight : 0)
                    total += Double(reps) * weight
                }
            }
        }
        return total
    }

    // MARK: - Expiration Check

    /// Marks a challenge as expired if its end date has passed.
    public static func checkExpiration(
        challenge: Challenge,
        referenceDate: Date = Date()
    ) -> Challenge {
        guard challenge.status == .active,
              let endDate = challenge.endDate,
              referenceDate > endDate else {
            return challenge
        }
        var updated = challenge
        updated.status = .expired
        return updated
    }

    // MARK: - Workout Volume Extraction

    /// Extracts the relevant volume from a single workout for a challenge.
    public static func workoutVolume(
        workout: Workout,
        exerciseName: String? = nil
    ) -> Double {
        if let name = exerciseName {
            // Per-exercise: only count matching exercises
            return workout.exercises
                .filter { $0.name == name }
                .flatMap { $0.targetSets }
                .reduce(0.0) { sum, set in
                    let reps = set.actualReps ?? set.reps
                    let weight = (set.completedWeight ?? 0) > 0
                        ? set.completedWeight!
                        : (set.prescribedWeight > 0 ? set.prescribedWeight : 0)
                    return sum + Double(reps) * weight
                }
        }
        return workout.totalVolume
    }
}

File: android/core/src/main/java/com/sundeefundee/core/domain/challenge/ChallengeEngine.kt
package com.sundeefundee.core.domain.challenge

import com.sundeefundee.core.model.Challenge
import com.sundeefundee.core.model.ChallengeProgress
import com.sundeefundee.core.model.ChallengeStatus
import com.sundeefundee.core.model.ChallengeTier
import com.sundeefundee.core.model.ChallengeType
import com.sundeefundee.core.model.Workout
import kotlinx.datetime.Clock
import kotlinx.datetime.TimeZone
import kotlinx.datetime.todayIn

// MARK: - ChallengeEngine

/// Pure domain logic for weight volume challenges.
///
/// All functions are side-effect-free — they accept data
/// and return data without touching any persistence layer.
object ChallengeEngine {

    // MARK: - Default Tiers

    /// Standard lifetime challenge tiers: Bronze (250K), Silver (500K), Gold (1M).
    fun defaultLifetimeTiers(): List<ChallengeTier> = listOf(
        ChallengeTier(name = "Bronze", targetVolumeLbs = 250_000.0, ordinal = 0),
        ChallengeTier(name = "Silver", targetVolumeLbs = 500_000.0, ordinal = 1),
        ChallengeTier(name = "Gold", targetVolumeLbs = 1_000_000.0, ordinal = 2)
    )

    // MARK: - Challenge Creation

    /// Creates a lifetime volume challenge with default tiers.
    fun createLifetimeChallenge(): Challenge {
        val now = Clock.System.todayIn(TimeZone.currentSystemDefault()).toString()
        return Challenge(
            id = java.util.UUID.randomUUID().toString(),
            type = ChallengeType.LIFETIME_VOLUME,
            title = "Lifetime Volume",
            tiers = defaultLifetimeTiers(),
            challengeStartDate = now,
            dateCreated = now
        )
    }

    /// Creates a per-exercise volume challenge.
    fun createExerciseChallenge(
        exerciseName: String,
        tiers: List<ChallengeTier>
    ): Challenge {
        val now = Clock.System.todayIn(TimeZone.currentSystemDefault()).toString()
        return Challenge(
            id = java.util.UUID.randomUUID().toString(),
            type = ChallengeType.EXERCISE_VOLUME,
            title = "$exerciseName Volume",
            exerciseName = exerciseName,
            tiers = tiers,
            challengeStartDate = now,
            dateCreated = now
        )
    }

    /// Creates a custom time-boxed challenge.
    fun createCustomChallenge(
        title: String,
        targetVolumeLbs: Double,
        endDate: String? = null
    ): Challenge {
        val now = Clock.System.todayIn(TimeZone.currentSystemDefault()).toString()
        val tier = ChallengeTier(name = "Goal", targetVolumeLbs = targetVolumeLbs, ordinal = 0)
        return Challenge(
            id = java.util.UUID.randomUUID().toString(),
            type = ChallengeType.CUSTOM,
            title = title,
            tiers = listOf(tier),
            challengeStartDate = now,
            challengeEndDate = endDate,
            dateCreated = now
        )
    }

    // MARK: - Progress Computation

    /// Computes the current progress of a challenge.
    fun computeProgress(challenge: Challenge): ChallengeProgress {
        val tiers = challenge.tiers.sortedBy { it.ordinal }

        if (tiers.isEmpty()) {
            return ChallengeProgress(
                currentTierName = "\u2014",
                percentComplete = 0.0,
                volumeRemaining = 0.0,
                isFullyComplete = true
            )
        }

        // If all tiers completed
        if (challenge.currentTierIndex >= tiers.size || challenge.status == ChallengeStatus.COMPLETED) {
            val lastTier = tiers.last()
            return ChallengeProgress(
                currentTierName = lastTier.name,
                percentComplete = 1.0,
                volumeRemaining = 0.0,
                isFullyComplete = true
            )
        }

        val currentTier = tiers[challenge.currentTierIndex]
        val previousTarget = if (challenge.currentTierIndex > 0) {
            tiers[challenge.currentTierIndex - 1].targetVolumeLbs
        } else {
            0.0
        }

        val tierRange = currentTier.targetVolumeLbs - previousTarget
        val tierProgress = challenge.accumulatedVolumeLbs - previousTarget
        val percent = if (tierRange > 0) {
            (tierProgress / tierRange).coerceIn(0.0, 1.0)
        } else {
            1.0
        }
        val remaining = maxOf(currentTier.targetVolumeLbs - challenge.accumulatedVolumeLbs, 0.0)

        return ChallengeProgress(
            currentTierName = currentTier.name,
            percentComplete = percent,
            volumeRemaining = remaining,
            isFullyComplete = false
        )
    }

    // MARK: - Volume Update

    /// Updates a challenge with additional volume, advancing tiers as needed.
    ///
    /// @return Pair of (updated challenge, progress with justCompletedTier set if applicable).
    fun updateVolume(
        challenge: Challenge,
        additionalVolumeLbs: Double
    ): Pair<Challenge, ChallengeProgress> {
        var accumulated = challenge.accumulatedVolumeLbs + additionalVolumeLbs
        var tierIndex = challenge.currentTierIndex
        val tiers = challenge.tiers.sortedBy { it.ordinal }
        var justCompleted: ChallengeTier? = null

        // Advance through tiers
        while (tierIndex < tiers.size) {
            val tier = tiers[tierIndex]
            if (accumulated >= tier.targetVolumeLbs) {
                justCompleted = tier
                tierIndex++
            } else {
                break
            }
        }

        // Determine final status
        val status = if (tierIndex >= tiers.size) {
            ChallengeStatus.COMPLETED
        } else {
            challenge.status
        }

        val updated = challenge.copy(
            accumulatedVolumeLbs = accumulated,
            currentTierIndex = tierIndex,
            status = status
        )

        val progress = computeProgress(updated).copy(
            justCompletedTier = justCompleted
        )

        return Pair(updated, progress)
    }

    // MARK: - Retroactive Volume

    /// Computes total historical volume from workout history.
    ///
    /// @param workouts All completed workouts.
    /// @param exerciseName If provided, only counts volume from this exercise.
    /// @return Total volume in pounds.
    fun retroactiveVolume(
        workouts: List<Workout>,
        exerciseName: String? = null
    ): Double {
        var total = 0.0
        for (workout in workouts) {
            if (!workout.isComplete) continue
            for (exercise in workout.exercises) {
                if (exerciseName != null && exercise.name != exerciseName) continue
                for (set in exercise.targetSets) {
                    val reps = set.actualReps ?: set.reps
                    val weight = if ((set.completedWeight ?: 0.0) > 0) {
                        set.completedWeight!!
                    } else if (set.prescribedWeight > 0) {
                        set.prescribedWeight
                    } else {
                        0.0
                    }
                    total += reps.toDouble() * weight
                }
            }
        }
        return total
    }

    // MARK: - Expiration Check

    /// Marks a challenge as expired if its end date has passed.
    fun checkExpiration(
        challenge: Challenge,
        referenceDate: String // ISO8601
    ): Challenge {
        if (challenge.status != ChallengeStatus.ACTIVE) return challenge
        val endDate = challenge.challengeEndDate ?: return challenge
        if (referenceDate > endDate) {
            return challenge.copy(status = ChallengeStatus.EXPIRED)
        }
        return challenge
    }

    // MARK: - Workout Volume Extraction

    /// Extracts the relevant volume from a single workout for a challenge.
    fun workoutVolume(
        workout: Workout,
        exerciseName: String? = null
    ): Double {
        if (exerciseName != null) {
            // Per-exercise: only count matching exercises
            return workout.exercises
                .filter { it.name == exerciseName }
                .flatMap { it.targetSets }
                .fold(0.0) { sum, set ->
                    val reps = set.actualReps ?: set.reps
                    val weight = if ((set.completedWeight ?: 0.0) > 0) {
                        set.completedWeight!!
                    } else if (set.prescribedWeight > 0) {
                        set.prescribedWeight
                    } else {
                        0.0
                    }
                    sum + reps.toDouble() * weight
                }
        }
        return workout.totalVolume
    }
}
